redis7/redis.go

```diff
@@ -0,0 +1,316 @@
+package redis7
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"github.com/go-redis/redis/v9"
+	"github.com/nuts-foundation/go-stoabs"
+	"github.com/nuts-foundation/go-stoabs/util"
+	"github.com/sirupsen/logrus"
+	"strings"
+	"sync"
+)
+
+const resultCount = 1000
+
+var _ stoabs.KVStore = (*store)(nil)
+var _ stoabs.ReadTx = (*tx)(nil)
+var _ stoabs.WriteTx = (*tx)(nil)
+var _ stoabs.Reader = (*shelf)(nil)
+var _ stoabs.Writer = (*shelf)(nil)
+
+// CreateRedisStore connects to a Redis database server using the given options.
+func CreateRedisStore(clientOpts *redis.Options, opts ...stoabs.Option) (stoabs.KVStore, error) {
+	cfg := stoabs.Config{}
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+
+	client := redis.NewClient(clientOpts)
+	_, err := client.Ping(context.TODO()).Result()
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect to Redis database: %w", err)
+	}
+
+	result := &store{
+		client: client,
+		mux:    &sync.RWMutex{},
+	}
+
+	result.log = cfg.Log
+	if result.log == nil {
+		result.log = logrus.StandardLogger()
+	}
+	return result, nil
+}
+
+type store struct {
+	client *redis.Client
+	log    *logrus.Logger
+	mux    *sync.RWMutex
+}
+
+func (s *store) Close(ctx context.Context) error {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	if s.client == nil {
+		// already closed
+		return nil
+	}
+	err := util.CallWithTimeout(ctx, s.client.Close, func() {
+		s.log.Error("Closing of Redis client timed out")
+	})
+	s.client = nil
+	return err
+}
+
+func (s *store) Write(fn func(stoabs.WriteTx) error, opts ...stoabs.TxOption) error {
```

Comment:
Redis doesn't have transactions so a parallel write to the same key will be undeterministic. The only way I see to prevent this is to add a mutex.

For example, when adding a transaction to the DAG, the clocks shelf will add the tx hash to the 'LC' key. The value for this key is a list. Two parallel adds will probably result in only adding one of the two hashes to the key.

Given previous example, Redis has a special operation for that but that operation must then also be supported on the writer interface. That might be a bit too much for now..

Reply:
Added a `TxOption` that acquires a global (write) lock on the store. Support in Redis through a library `github/redsync` (although this PR uses a fork because my PR to support Redis 9 on it is pending), BBolt always locks.